### DISCUSSION TITLE
The spider has a black paw

### DISCUSSION THREAD

**Discussion by stla** [pyvista/pyvista#2045] [Show and tell]
Hello,

  I found something strange with the [spider](https://docs.pyvista.org/api/examples/_autosummary/pyvista.examples.downloads.download_spider.html#). I import it in Python, then I save the points, the faces and the normals in CSV files, and then I read these CSV files in R, and I reconstruct the mesh (with the **rgl** package, a wrapper of OpenGL). Then the spider has a black paw:

![spider_black_paw](https://user-images.githubusercontent.com/4466543/150308118-51659d50-0481-446b-b79d-cf0d7570bd2b.png)

It is well known that a mesh is rendered in black when its normals have the opposite sign of the sign they should have. Now, I use the **Rvcg** package, a wrapper of the C++ library **VCGLIB**. It has a function to "repair" a mesh (which works only when the mesh is not too screwed up). I apply this function to the spider mesh, and then the color of the black paw becomes correct:

![spider_clean](https://user-images.githubusercontent.com/4466543/150308947-e27bdbd7-162e-477c-afd1-6133d425684f.png)

**Comment by stla** [comment]
In fact, even with PyVista there is some blackness on this paw when one enables the `smooth_shading`, but this is only a black line:

![spider_pyvista](https://user-images.githubusercontent.com/4466543/150310500-48312307-babe-4e2f-8095-30055c9a4380.png)

**Comment by adeak** [comment]
Good catch: it seems that on half of that leg the normals are reversed, so they overall point to the same side of that leg (rather than radially around it like on every other appendage):

![tmp](https://user-images.githubusercontent.com/17914410/150330449-3c54f7f1-ac76-4850-b310-5a69a7880b8d.png)

I don't know how we could fix that.

**Reply by adeak** [reply]
`dataset.compute_normals(auto_orient_normals=True)` fixes the normals on the leg... and introduces the same error on the other foreleg :confused:

**Reply by adeak** [reply]
This is extra confusing to me because the spider is _almost_ manifold, and its two boundary edges are in the middle of its body (in the cephalothorax). So I'd expect auto orientation to work, or at least work differently. But I don't know how it works.

**Comment by stla** [comment]
Works like a charm in R, with the `vcgClean` function of the `Rvcg` package:

```r
library(rgl)

pts <- read.csv("spider_points.csv", header = FALSE)
normals <- read.csv("spider_normals.csv", header = FALSE)
faces <- read.csv("spider_faces.csv", header = FALSE)[, 1L]
faces <- faces[seq_along(faces) %% 4L != 1L] + 1L
faces <- matrix(faces, nrow = 3L)

mesh0 <- tmesh3d(t(as.matrix(pts)), faces, normals = as.matrix(normals))
mesh <- Rvcg::vcgClean(mesh0, sel = 7)
shade3d(mesh, color = "tan")
Morpho::plotNormals(mesh, length = 0.3, col = "cyan")
```

![spider_normals](https://user-images.githubusercontent.com/4466543/150340877-dfc8d941-7fbb-4299-a858-e720993b4551.png)

**Reply by akaszynski** [reply]
Could you write this out as a ply, upload here, or overwrite the existing example file.

**Reply by stla** [reply]
@akaszynski Sure. Here are the ply files, in binary and non-binary types.
[spider_ply.zip](https://github.com/pyvista/pyvista/files/7911953/spider_ply.zip)
